src/core/agent/services/connectionService.ts

```diff
@@ -299,21 +306,52 @@ class ConnectionService extends AgentService {
   async getConnectionHistoryById(
     connectionId: string
   ): Promise<ConnectionHistoryItem[]> {
-    let histories: ConnectionHistoryItem[] = [];
     const credentialRecords =
       await this.credentialStorage.getCredentialMetadataByConnectionId(
         connectionId
       );
-    histories = histories.concat(
-      credentialRecords.map((record) => {
-        return {
-          type: ConnectionHistoryType.CREDENTIAL_ACCEPTED,
-          timestamp: record.createdAt.toISOString(),
-          credentialType: record.credentialType,
+    const credentialStatistics: {
+      [key: string]: { name: string; count: number };
+    } = {};
+    credentialRecords.forEach((record) => {
+      if (!credentialStatistics[record.id.replace("metadata:", "")]) {
+        credentialStatistics[record.id.replace("metadata:", "")] = {
+          name: record.credentialType,
+          count: 0,
         };
-      })
+      }
+    });
+
+    const linkedIpexMessages = await this.getLinkedIPEXMessageByConnectionId(
+      connectionId
     );
-    return histories;
+
+    const requestMessages = linkedIpexMessages
+      .sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime())
+      .map((messageRecord) => {
+        const credentialStatistic =
+          credentialStatistics[messageRecord.content.exn.e.acdc.d];
+        let historyType: ConnectionHistoryType;
+        const route = messageRecord.content.exn.r;
+        if (
+          route === ExchangeRoute.IpexGrant &&
+          credentialStatistic.count === 0
+        ) {
+          historyType = ConnectionHistoryType.CREDENTIAL_ISSUANCE;
+        } else if (route === ExchangeRoute.IpexApply) {
+          historyType = ConnectionHistoryType.CREDENTIAL_PRESENT;
```

Comment:
Actually, a presentation is also `IpexGrant` - if someone is presenting to us, we haven't done it yet though.

If we receive `IpexApply` - it means someone is requesting us to present